File: sections/14_core_patterns_mixed_contest/problems/u_circular_fuel_start/solution.py
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    n = data[0]
    fuel = data[1:n + 1]
    cost = data[n + 1:]
    total = tank = candidate = 0
    for i, (received, spent) in enumerate(zip(fuel, cost)):
        gain = received - spent
        total += gain
        tank += gain
        if tank < 0:
            candidate = i + 1
            tank = 0
    print(-1 if total < 0 else candidate)


if __name__ == "__main__":
    main()

File: sections/14_core_patterns_mixed_contest/problems/u_circular_fuel_start/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n;
    cin >> n;
    vector<long long> fuel(n), cost(n);
    for (long long &value : fuel) cin >> value;
    for (long long &value : cost) cin >> value;
    long long total = 0, tank = 0;
    int candidate = 0;
    for (int i = 0; i < n; ++i) {
        long long gain = fuel[i] - cost[i];
        total += gain;
        tank += gain;
        if (tank < 0) {
            candidate = i + 1;
            tank = 0;
        }
    }
    cout << (total < 0 ? -1 : candidate) << '\n';
}
